src/classes/UTIL_DescribeProvider.cls

```diff
@@ -0,0 +1,51 @@
+/*
+    Copyright (c) 2019 Salesforce.org
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Utilities
+* @description This class facilitates the use of dependency injection to mock calls to
+* instance methods in the UTIL_Describe class within tests.
+*/
+public with sharing class UTIL_DescribeProvider {
```

Comment:
What's the benefit of having a separate provider class, as opposed to having utilDescribeInstance being a static member variable of UTIL_Describe itself and supporting these two methods under that class? I'm wondering if only because we haven't included separate providers in our repo as yet, and I don't want to introduce the pattern into our code base lightly if it's not one we at least want to move towards. That, and we have more than 1000 class files in our repo already!

Is there any downside to just moving the guts of this class into UTIL_Describe (and doing similar transfer for UTIL_Permissions) and basically allowing those classes to serve as providers for themselves?

Reply:
Yeah, that's a good catch - in this case it probably doesn't provide any specific value.  If the UTIL_Describe or UTIL_Permissions classes were implementations of interfaces, then having a separate provider class would enable complete switching of implementations down the road - but since they are not it makes sense to have the provider function inside the class itself.  I'll switch it up on a follow-up commit!